File: ch4/service/ch4_service_cpp/src/self_service.cpp
// Copyright (c) 2022 Homalozoa
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <memory>
#include <string>
#include <thread>
#include <utility>

#include "rclcpp/rclcpp.hpp"

class ServerNode : public rclcpp::Node
{
public:
  explicit ServerNode(const std::string & node_name)
  : Node(node_name)
  {
    using namespace std::chrono_literals;
    auto logtimer_callback =
      [&]() -> void {
        RCLCPP_INFO_STREAM(this->get_logger(), "Hello world");
      };
    timer_ = this->create_wall_timer(500ms, logtimer_callback);
    server_ = this->create_service<rcl_interfaces::srv::GetParameters>(
      "get_para",
      std::bind(&ServerNode::service_callback, this, std::placeholders::_1, std::placeholders::_2));
  }

private:
  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::Service<rcl_interfaces::srv::GetParameters>::SharedPtr server_;
  void service_callback(
    const std::shared_ptr<rcl_interfaces::srv::GetParameters::Request> request,
    std::shared_ptr<rcl_interfaces::srv::GetParameters::Response> response)
  {
    rcl_interfaces::msg::Parameter para_;
    para_.value.bool_value = !request->names.empty();
    response->values.push_back(para_.value);
    // std::this_thread::sleep_for(std::chrono::seconds(3));
    RCLCPP_INFO_STREAM(
      this->get_logger(),
      "Request: " << request->names[0]);
  }
};

class ClientNode : public rclcpp::Node
{
public:
  explicit ClientNode(const std::string & node_name)
  : Node(node_name)
  {
    using namespace std::chrono_literals;
    client_ = this->create_client<rcl_interfaces::srv::GetParameters>("get_para");
    auto clientimer_callback =
      [&]() -> void {
        auto req = std::make_unique<rcl_interfaces::srv::GetParameters::Request>();
        req->names.push_back(std::to_string(this->get_clock()->now().seconds()));
        // if (client_->service_is_ready()) {
        RCLCPP_INFO_STREAM(this->get_logger(), "Ready to send req");
        client_->async_send_request(std::move(req));
        RCLCPP_INFO_STREAM(this->get_logger(), "Sent req");
        // }
      };
    clientimer_ = this->create_wall_timer(500ms, clientimer_callback);
    auto logtimer_callback =
      [&]() -> void {
        RCLCPP_INFO_STREAM(this->get_logger(), "Hello earth");
      };
    logtimer_ = this->create_wall_timer(500ms, logtimer_callback);
  }

private:
  rclcpp::TimerBase::SharedPtr clientimer_;
  rclcpp::TimerBase::SharedPtr logtimer_;
  rclcpp::Client<rcl_interfaces::srv::GetParameters>::SharedPtr client_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto srv_node_ = std::make_shared<ServerNode>("srv_server");
  auto clt_node_ = std::make_shared<ClientNode>("srv_client");
  rclcpp::executors::SingleThreadedExecutor executor_;
  // rclcpp::executors::MultiThreadedExecutor executor_;

  executor_.add_node(srv_node_);
  executor_.add_node(clt_node_);
  executor_.spin();

  rclcpp::shutdown();
  return 0;
}

File: ch4/service/ch4_service_py/ch4_service_py/self_service.py
# Copyright (c) 2022 Homalozoa
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import rclpy
import rclpy.executors

from rclpy.node import Node
from rcl_interfaces.msg import Parameter
from rcl_interfaces.srv import GetParameters


class ServerNode(Node):

    def __init__(self, name):
        super().__init__(name)
        self.callback_group = rclpy.callback_groups.ReentrantCallbackGroup()
        self.serv_ = self.create_service(GetParameters, 'get_para', self.srv_callback,
                                         callback_group=self.callback_group)

    def srv_callback(self, request, response):
        para_ = Parameter()
        para_.value.bool_value = bool(request.names)
        self.get_logger().info('Request: ' + request.names[0])
        response.values.append(para_.value)
        return response


class ClientNode(Node):

    def __init__(self, name):
        super().__init__(name)
        self.callback_group = rclpy.callback_groups.ReentrantCallbackGroup()
        self.client_ = self.create_client(GetParameters, 'get_para')
        self.create_timer(0.5, self.clientimer_callback, self.callback_group)

    def clientimer_callback(self):
        request_ = GetParameters.Request()
        now_sec, _ = self.get_clock().now().seconds_nanoseconds()
        self.get_logger().info('Ready to send req')
        request_.names.append(str(now_sec))
        future_ = self.client_.call(request_)
        self.get_logger().info('Sent req')

        response_value_ = future_.values[0]
        if bool(response_value_.bool_value):
            self.get_logger().info('Got response succeed')


def main(args=None):
    rclpy.init(args=args)
    srv_node = ServerNode('service_server')
    cli_node = ClientNode('service_client')
    executor = rclpy.executors.MultiThreadedExecutor()
    executor.add_node(srv_node)
    executor.add_node(cli_node)
    executor.spin()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
